### PR TITLE
Support the way to migrate whole databse with `--dump {PATH}` argument (+`to_ddl()`)

### PR DESCRIPTION
**Pull request by devgony**
### Dump whole schemas and data by generating SQL with `--dump {PATH}` option
```sh
gluesql --path ~/gluedata --dump ./dump.sql
```
```sql
-- dump.sql
CREATE TABLE User (id INT, name TEXT);
CREATE INDEX User_id ON User (id);
..
INSERT INTO User VALUES (1, 'Foo'), (2, 'Bar') ..
..
```
### Import database
```sh
gluesql --path ~/newdata --execute ./dump.sql
```
### Example of dump.sql
```sql
CREATE TABLE Test (N INT NOT NULL);
CREATE INDEX Test_N ON Test (N);
INSERT INTO Test VALUES (1, "b"), (2, "b"), (3, "b"), (4, "b"), (5, "b"), (6, "b"), (7, "b"), (8, "b"), (9, "b"), (10, "b"), (11, "b"), (12, "b"), (13, "b"), (14, "b"), (15, "b"), (16, "b"), (17, "b"), (18, "b"), (19, "b"), (20, "b"), (21, "b"), (22, "b"), (23, "b"), (24, "b"), (25, "b"), (26, "b"), (27, "b"), (28, "b"), (29, "b"), (30, "b"), (31, "b"), (32, "b"), (33, "b"), (34, "b"), (35, "b"), (36, "b"), (37, "b"), (38, "b"), (39, "b"), (40, "b"), (41, "b"), (42, "b"), (43, "b"), (44, "b"), (45, "b"), (46, "b"), (47, "b"), (48, "b"), (49, "b"), (50, "b"), (51, "b"), (52, "b"), (53, "b"), (54, "b"), (55, "b"), (56, "b"), (57, "b"), (58, "b"), (59, "b"), (60, "b"), (61, "b"), (62, "b"), (63, "b"), (64, "b"), (65, "b"), (66, "b"), (67, "b"), (68, "b"), (69, "b"), (70, "b"), (71, "b"), (72, "b"), (73, "b"), (74, "b"), (75, "b"), (76, "b"), (77, "b"), (78, "b"), (79, "b"), (80, "b"), (81, "b"), (82, "b"), (83, "b"), (84, "b"), (85, "b"), (86, "b"), (87, "b"), (88, "b"), (89, "b"), (90, "b"), (91, "b"), (92, "b"), (93, "b"), (94, "b"), (95, "b"), (96, "b"), (97, "b"), (98, "b"), (99, "b"), (100, "b");
INSERT INTO Test VALUES (101, "b");

CREATE TABLE User (id INT NOT NULL, name TEXT NOT NULL);
CREATE INDEX User_id ON User (id);
INSERT INTO User VALUES (1, "a"), (2, "a"), (3, "a"), (4, "a"), (5, "a"), (6, "a"), (7, "a"), (8, "a"), (9, "a"), (10, "a"), (11, "a"), (12, "a"), (13, "a"), (14, "a"), (15, "a"), (16, "a"), (17, "a"), (18, "a"), (19, "a"), (20, "a"), (21, "a"), (22, "a"), (23, "a"), (24, "a"), (25, "a"), (26, "a"), (27, "a"), (28, "a"), (29, "a"), (30, "a"), (31, "a"), (32, "a"), (33, "a"), (34, "a"), (35, "a"), (36, "a"), (37, "a"), (38, "a"), (39, "a"), (40, "a"), (41, "a"), (42, "a"), (43, "a"), (44, "a"), (45, "a"), (46, "a"), (47, "a"), (48, "a"), (49, "a"), (50, "a"), (51, "a"), (52, "a"), (53, "a"), (54, "a"), (55, "a"), (56, "a"), (57, "a"), (58, "a"), (59, "a"), (60, "a"), (61, "a"), (62, "a"), (63, "a"), (64, "a"), (65, "a"), (66, "a"), (67, "a"), (68, "a"), (69, "a"), (70, "a"), (71, "a"), (72, "a"), (73, "a"), (74, "a"), (75, "a"), (76, "a"), (77, "a"), (78, "a"), (79, "a"), (80, "a"), (81, "a"), (82, "a"), (83, "a"), (84, "a"), (85, "a"), (86, "a"), (87, "a"), (88, "a"), (89, "a"), (90, "a"), (91, "a"), (92, "a"), (93, "a"), (94, "a"), (95, "a"), (96, "a"), (97, "a"), (98, "a"), (99, "a"), (100, "a");
INSERT INTO User VALUES (101, "a"), (102, "a");
```

## Todo
- [x] implement to_ddl on `Schema`
- [x] implement `--dump {PATH}`
  - [x] DDL
  - [x] INSERT Statement
- [x] write migration guide on README.md
- [x] handle Value::{Map, List}
- [x] remove unwrap

## Next
- modify NOT NULL to NULL, add nullable field on ColumnDef
- function::GET_DDL

### CODE REVIEW

File: cli/src/lib.rs [hunk 1]
```diff
@@ -24,14 +31,36 @@ struct Args {
...
     let args = Args::parse();
 
     if let Some(path) = args.path {
         let path = path.as_path().to_str().expect("wrong path");
 
+        if let Some(dump) = args.dump {
+            block_on(async {
+                let storage = SledStorage::new(path).expect("failed to load sled-storage");
+                let schemas = storage.fetch_all_schemas().await?;
+                println!("{schemas:?}");
```
Comment (devgony) [inline cli/src/lib.rs:50]: It's so weired fetch_all_schemas returns empty Vector even if there is Table and Index for sure..

File: core/src/ast/ast_literal.rs [hunk 2]
```diff
@@ -26,6 +28,43 @@ impl ToSql for AstLiteral {
...
+            // Value::Timestamp(v) => {
+            //     AstLiteral::QuotedString(DateTime::<Utc>::from_utc(v, Utc).to_string().into())
+            // }
+            Value::Time(v) => AstLiteral::QuotedString(v.to_string()),
+            Value::Interval(v) => AstLiteral::QuotedString(v.into()),
+            Value::Uuid(v) => {
+                AstLiteral::QuotedString(Uuid::from_u128(v).hyphenated().to_string().into())
+            }
+            Value::Map(v) => todo!(),
+            Value::List(v) => todo!(),
```
Comment (devgony) [inline core/src/ast/ast_literal.rs:60]: We have to decide how to show Map and List with Literal

File: core/src/ast/ast_literal.rs [hunk 3]
```diff
@@ -26,6 +28,43 @@ impl ToSql for AstLiteral {
...
+            Value::F64(v) => AstLiteral::Number(BigDecimal::from_f64(v).unwrap()),
+            Value::Decimal(v) => {
+                AstLiteral::Number(BigDecimal::from_f64(v.try_into().unwrap()).unwrap())
+            }
+            Value::Str(v) => AstLiteral::QuotedString(v),
+            Value::Bytea(v) => AstLiteral::HexString(hex::encode(v)),
+            Value::Date(v) => AstLiteral::QuotedString(v.to_string()),
+            Value::Timestamp(v) => AstLiteral::QuotedString(v.to_string()),
+            // Value::Timestamp(v) => {
+            //     AstLiteral::QuotedString(DateTime::<Utc>::from_utc(v, Utc).to_string().into())
```
Comment (devgony) [inline core/src/ast/ast_literal.rs:52]: 🤔 should we add Utc here?

File: cli/src/lib.rs [hunk 4]
```diff
@@ -24,14 +32,79 @@ struct Args {
...
+                                        order_by: Vec::new(),
+                                        limit: None,
+                                        offset: None,
+                                    },
+                                }
+                                .to_sql();
+
+                                writeln!(&file, "{}", insert_statement)
+                                    .map_err(ready)
+                                    .unwrap();
```
Comment (devgony) [inline cli/src/lib.rs:91]: So hard to merge TryChunkError + I/O Error + gluesqlError here but eventually we will just unwrap at main..

File: core/src/ast/ast_literal.rs (hunk 2 above)
Comment (panarch) [reply core/src/ast/ast_literal.rs]: Hmm... this really looks interesting.
What we need is generate SQL from AST so you are trying to convert `Value` to `AstLiteral`.

Comment (devgony) [reply core/src/ast/ast_literal.rs]: Yes.. It is the first case against our parsing direction

Comment (panarch) [reply core/src/ast/ast_literal.rs]: Currently thinking about
expanding `Expr` to contain `Value`, then we can impl `ToSql` for `Value`.

File: core/src/data/schema.rs [hunk 5]
```diff
@@ -20,13 +20,46 @@ pub struct SchemaIndex {
...
+
+        let create_indexes =
+            indexes
+                .iter()
+                .fold("".to_owned(), |acc, SchemaIndex { name, expr, .. }| {
+                    let expr = expr.to_sql();
+                    let table_name = &table_name;
+
+                    format!("{acc}CREATE INDEX {name} ON {table_name} ({expr});\n")
+                });
```
Comment (panarch) [inline core/src/data/schema.rs:47]: nit suggestion.
```rust
let create_indexes = indexes
    .iter()
    .map(|SchemaIndex { name, expr, .. }| {
        let expr = expr.to_sql();
        let table_name = &table_name;

        format!("CREATE INDEX {name} ON {table_name} ({expr});")
    })
    .collect::<Vec<_>>()
    .join("\n");
```

File: core/src/data/schema.rs [hunk 6]
```diff
@@ -20,13 +20,46 @@ pub struct SchemaIndex {
...
+            if_not_exists: false,
+            name: table_name,
+            columns,
+            source: None,
+        }
+        .to_sql();
+
+        format!("{create_table}\n{create_indexes}")
+            .trim_end()
+            .to_owned()
```
Comment (panarch) [inline core/src/data/schema.rs:59]: Then, we can remove these lines

Comment (panarch) [reply core/src/data/schema.rs:59]:
```rust
vec![create_table, create_indexes].join("\n")
```

**Review by panarch — APPROVED** [review]
Awesome 👍 👍 
All the codes and documentation look all nice.
Thanks a lot!

**Pull request merged**